### PR TITLE
Replace raiseEndN's with general raiseEnd

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
While going through the code, this bit jumped out and it was too tempting to try to improve it. This approach might very well already have been considered, so feel free to close this in that case.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you changed any Haskell files:
    - [x] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: freer-extras/src/Control/Monad/Freer/Extras/Modify.hs [hunk 1]
```diff
@@ -91,95 +59,14 @@ inserting effects at the *end* of the list is tricky.
...
-weakenEnd17 :: forall effs a b c d e f g h i j k l m n o p q. Union '[a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q] ~> Union (a ': b ': c ': d ': e ': f ': g ': h ': i ': j ': k ': l ': m ': n ': o ': p ': q ': effs)
-weakenEnd17 u = case decomp u of
-    Left u' -> weaken $ weakenEnd16 u'
-    Right t -> inj t
-
-weakenEnd18 :: forall effs a b c d e f g h i j k l m n o p q r. Union '[a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r] ~> Union (a ': b ': c ': d ': e ': f ': g ': h ': i ': j ': k ': l ': m ': n ': o ': p ': q ': r ': effs)
-weakenEnd18 u = case decomp u of
-    Left u' -> weaken $ weakenEnd17 u'
-    Right t -> inj t
+class CanWeakenEnd as effs where
```
Comment (michaelpj) [inline freer-extras/src/Control/Monad/Freer/Extras/Modify.hs:62]: This is actually a very straightforward implementation, not sure why I didn't come up with it. LGTM!

**Review by michaelpj — APPROVED** [review]
Literal tears of joy over here.

**Comment by michaelpj** [comment]
I think I got stuck trying to do what we talked about in that issue and implement a fully generic `RaiseBehind` class. This approach seems much better.

File: freer-extras/src/Control/Monad/Freer/Extras/Modify.hs (hunk 1 above)
Comment (sjoerdvisscher) [reply freer-extras/src/Control/Monad/Freer/Extras/Modify.hs:62]: Yeah, I started out with something much more complicated (thinking you'd need to do something like `raiseEnd @7`) and once I got something working it turned out it kept working after stripping stuff out bit by bit.

**Comment by michaelpj** [comment]
Oh yeah, please do fix the comment.

**Comment by sjoerdvisscher** [comment]
I think everything in the note is still true. But I should add a bit about that there's a double cons in the second instance to prevent overlapping instances.

**Review by effectfully — APPROVED** [review]

File: freer-extras/src/Control/Monad/Freer/Extras/Modify.hs [hunk 2]
```diff
@@ -91,95 +59,14 @@ inserting effects at the *end* of the list is tricky.
...
+    weakenEnd :: Union as ~> Union effs
+instance a ~ e => CanWeakenEnd '[a] (e ': effs) where
+    weakenEnd u = inj (extract u)
+instance (a ~ e, CanWeakenEnd (b ': as) (f ': effs)) => CanWeakenEnd (a ': b ': as) (e ': f ': effs) where
+    weakenEnd u = case decomp u of
+        Left u' -> weaken $ weakenEnd u'
+        Right t -> inj t
```
[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply freer-extras/src/Control/Monad/Freer/Extras/Modify.hs:73]: Sounds good, let me try.

[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply freer-extras/src/Control/Monad/Freer/Extras/Modify.hs:73]: Yes nice! On this example

```haskell
raiseEnd2 :: Eff '[a, b] ~> Eff _
raiseEnd2 = raiseEnd
```

I now get `Found type wildcard ‘_’ standing for ‘a : b : effs' :: [* -> *]’`, which didn't happen before.

**Comment by michaelpj** [comment]
Nice.

**Pull request merged**